pyiron_base/generic/datacontainer.py

```diff
@@ -510,6 +510,43 @@ def get(self, key, default=None, create=False):
         else:
             return super().get(key, default=default)
 
+    def search_tree(self, key, fail=True):
+        """
+        Search for ``key`` in the Container hierarchy.
+
+        This should be used if there is only one item in the hierarchy.
+        If the key appears multiple times, there is no control over which
+        item is found.
+
+        Args:
+            key (str):          the key to look for
+            fail (bool):        what to do if key is not found.
+                                True  => raise IndexError
+                                False => return None
+        Raise:
+            TypeError:  if key is not str
+            IndexError: if key is not found and fail is True.
+
+        Returns:
+            object: element at ``key``
+        """
+
+        if not isinstance (key,str):
+            raise TypeError ("Cannot search for non-string key.")
+
+        # search within current level
+        if (key in self): return self.get(key)
+
+        # descend into subgroups
+        for it in self.groups ():
+            pick = search_tree(self[it], key, False)
+            if (not pick is None): return pick
```

Comment:
Did you add `fail` just for this check or is that something we want to expose to users?  If the former I think always trowing an exception/catching it and not having the argument feels a bit cleaner.

Reply:
I do a recursion here. Inside the recursion, not finding the key is not an error - it just means I didn't search the right part yet. Only if the key is nowhere to be found, there is a true problem.
There is different alternatives to the current solution:
- simply return None if key is not found (no fail needed). Disadvantage - you don't get a very useful error message if you had expected the key to be found. 
- always raise an IndexError, but catch it inside the recursion. I had once learned that one should use exceptions for exceptional cases, because exception handling is not designed to be performant. I don't know if this is true for python. That's why I decided against this solution

I went for the current solution because it allows the user to decide whether or not he/she prefers an error, or None return value.

Reply:
renamed "fail" into "fail_if_not_found" to make more clear what this arg is about.